### PR TITLE
misc: make ArrayAttr covariant by using tuple instead of list

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

**Review by webmiche — COMMENTED** [review]
Nice, I like the idea. Once pyright is happy, I am fine with merging!

**Review by georgebisbas — APPROVED** [review]

**Comment by superlopuh** [comment]
Please disregard pyright issues, they're fixed in another branch. We're still at 0 errors with the Pyright CI that knows to ignore builtin.py

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -1,9 +1,10 @@
 from __future__ import annotations
+from abc import ABC, abstractmethod
```
[the remark this answers is not in the document]
Comment (math-fehr) [reply xdsl/dialects/builtin.py:2]: Can we remove this error at least? I don't see why not removing it now

File: xdsl/dialects/builtin.py [hunk 2]
```diff
@@ -85,7 +90,8 @@ def verify(self) -> None:
     @staticmethod
     @deprecated_constructor
     def from_list(data: List[_ArrayAttrT]) -> ArrayAttr[_ArrayAttrT]:
-        return ArrayAttr(data)
+        d: Iterable[_ArrayAttrT] = data
+        return ArrayAttr[_ArrayAttrT](d)
```
Comment (math-fehr) [inline xdsl/dialects/builtin.py:94]: Is it still needed? I don't remember why I wrote that initially

**Pull request merged**